### ISSUE TITLE
Using `cache.modify` in a subscription with Typescript.

### ISSUE THREAD

**Issue by hickscorp** [apollographql/apollo-client#11623]
We have a subscription that notifies us when something new is added to a connection.
In our React component tree, we have something like this:

```typescript
export const Subscription = ({ id, bearerToken }: ObserverProps) => {
  const { cache } = useApolloClient();
  const sub = useSubscription(SUBSCRIPTION, {
    variables: { id, bearerToken },
  });
  useEffect(() => {
    if (sub.loading || !sub.data) return;
    cache.modify({
      id: cache.identify(sub.data.conversationStepCreated.conversation),
      fields: {
        steps: (existingRefs) => {
          const { edges } = existingRefs;
        },
      },
    });
  }, [cache, sub]);
  return null;
};
```

When the subscription ticks, we receive a `Step` entity, which should be added to its parent `Conversation.steps` connection in the cache so that UI updates automatically.

I am failing to understand how to use `cache.modify` to do so:
- While having type safety,
- Generally speaking - what do I append to what?

EDIT: We've tried fetching the "last one step" from the conversation upon the subscription firing - but it seems to be killing our cache alltogether and leaves it with just one entry.

```typescript
export const Subscription = ({ id, bearerToken }: ObserverProps) => {
  const sub = useSubscription(SUBSCRIPTION, {
    variables: { id, bearerToken },
  });

  const [stepsQuery] = useLazyQuery(CONVERSATION_STEPS_QUERY, {
    fetchPolicy: "network-only",
  });

  useEffect(() => {
    if (sub.loading || !sub.data) return;
    stepsQuery({ variables: { id } });
  }, [id, stepsQuery, sub.loading, sub.data]);
  return null;
};
```

**Comment by hickscorp** [comment]
Good morning and thank you @phryneas .

- We have a parent entity - `Conversation` that has a connection to edges of type `Step`.
- We have a page listing `Steps` inside a `Conversation`. It has some logic to `fetchMore` by shifting over the `PageInfo` cursor.
- We have a subscription that triggers when a new `Step` becomes part of a `Conversation` remotelly, maybe created by another user.
- Upon this subscription firing, we receive the `Step` that was added, and we want to add it to the Apollo cache - the one being used behind the scenes, being primed and filled by the page listing the `Step` entities.

So from what we understood, a good way to do so would be to use `cache.modify` on the `Connection` edges to append the `Step` received by the subscription to the edges that were already fetched.
What I don't understand is how to do this in a "type safety" fashion, or even how to approach the data structures that should be involved in pushing new elements to the cache. It'd be nice to have a working example of such scenario.

In response to your comments:
- We don't use `useLazyQuery` lightly in `useEffect`. Sometimes it's necessary - and with the right dependencies we never had problems with that. I'm curious why you flag this though, would you care to ellaborate please? In our case, using `useQuery` along with the `skip` flag isn't an option, because we can't have this `useQuery` declared upfront with the variables - as our component is mounted with these variables being optional - and therefore can't compile without them being set to a value. This is what we want from the strong typing guarantees at compile-time - and instead we want the query to only fire when the variables become set. We also don't want to start getting into "nested nested components" hell so that we can have a conditionally rendered component that `useQuery` internally, it essentially results in the same internal logic with more complexity in our code.
- Noted, very good to know. Thanks!

**Comment by phryneas** [comment]
> This is what we want from the strong typing guarantees at compile-time - and instead we want the query to only fire when the variables become set.

Understood. Personally, I would prefer the better "runtime code" over the type safety here, but that's very opinionated.
In the future, you'll be able to do this in a more typesafe matter using `skipToken`, which is already available with `useSuspenseQuery`.

---

Generally, you're in a bit of an edge case situation regarding type safety - usually, you'd do something like this:
```ts
      cache.modify<EntityType>({
        id: ...,
        fields: {
          fieldName(valueOrRef) {
            return differentValue
          }
        }
      });
```
You can see more usage examples in our tests, e.g. at https://github.com/apollographql/apollo-client/blob/9dc45bb8d13ebb486dd1cb9963e1e0d9787d984d/src/cache/core/__tests__/cache.ts#L348-L383

Your problem here is that your have deeper nesting here - which the cache will *try* to normalize, and if it's successful, nested fields like `steps.edges` here will only be another reference.

So depending on how your cache can be normalized, you'd have to actually call `cache.modify` for the instance stored in `sub.data.conversationStepCreated.conversation.steps` instead, with the appropriate type for that.

**Comment by hickscorp** [comment]
Thansk a lot @phryneas this is helpful.

I'm trying to simplify the codebase to give a clearer example, as it seems that the question was misunderstood. Will come back to you soon guys :)

**Comment by hickscorp** [comment]
Ok, so I think I found a way to ask way, way more simply.

Imagine a `Post` object with a `connection` field `comments` of type `Comment`.
The client lists these comments by performing a "one shot" query to the post.
At the same time, it subscribes to a subscription `CommentCreated(postId: ...)` which gives back a `Comment`.

I would imagine that the Apollo client cache would be configured with something like this:
```typescript
new InMemoryCache({
    fragments: createFragmentRegistry(),
    typePolicies: {
      Post: {
        fields: {
          comments: relayStylePagination(),
        },
      },
    }
```

Cool.

The more simple question is - using TypeScript - when the `CommentCreated` subscription fires, how do you append that new post to the existing list **in the cache**, so that when the user changes page and comes back to the same post, it's persisted?

**Comment by hickscorp** [comment]
We've been successful with modifying the cache, with something like this - this being the function that is called when the subscription fires:

```typescript
  const onNewComment = (comment: DetailedCommentFragment) => {
    cache.modify({
      id: cache.identify({
        __typename: "Post",
        id: comment.post.id,
      }),
      fields: {
        comments(existing) {
          return {
            ...existing,
            edges: [...existing.edges, { node: comment }],
          };
        },
      },
    });
```

The proble here is that `existing` is `any`. There are no guarantees at all, and it feels like blind luck that it works because it could in fact be a concrete `CommentConnection` but it could also be a `Reference`...
The moment we add a type hint to `cache.modify<Post>(...)` to our code, it breaks because we're not handling `Reference`... So we played around with `readField` but can't seem to succeed with finding the right types to use.
Any sample code that would help deal with the Relay types (connections, edges having a cursor etc and appending to them) could be useful here.

Thanks a lot!

**Comment by hickscorp** [comment]
Also, looking at `https://github.com/apollographql/apollo-client/blob/main/src/utilities/policies/pagination.ts#L94` it seems that this could do exactly what we're looking for - but we have no idea on how to use it after the subscription fires.

**Comment by hickscorp** [comment]
Ok - a bit of progress.
It seems that we're getting a bit closer to type safety, if we use `cache.updateQuery` instead of `cache.modify`.

Would this be a good way to achieve what we want? It seems to be working in a (very) controlled environment:

```typescript
  const onNewStep = (step: DetailedStepFragment) => {
    cache.updateQuery(
      {
        query: GQL.Conversation.StepsQuery,
        variables: { id: step.conversation.id, last: 8 },
      },
      (existing) => {
        if (existing?.node?.__typename !== "Conversation") return;
        const node = frag(GQL.Conversation.WithSteps, existing.node);
        node.steps.edges;
        return {
          ...existing,
          node: {
            ...node,
            steps: {
              ...node.steps,
              edges: [
                ...node.steps.edges,
                { __typename: "ConversationStepEdge", node: step },
              ],
            },
          },
        };
      }
    );
  };
```

**Comment by phryneas** [comment]
I'm sorry, I dropped the ball on this issue, just too much to do in too little time :(

Did your solution work out for you in the end?

Also, as a heads up, we recently released [Version 3.11](https://github.com/apollographql/apollo-client/releases/tag/v3.11.0) with a rewrite of `useSubscription`. It comes with an `ignoreResults` option, so you could set that to `true` and move your effect into the `onData` option callback. That would give you more control over component rerenders.
